# 01. Importing libraries

In [40]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

## set path

In [41]:
path = r'/Users/marcela/Achievement 4'

In [42]:
path

'/Users/marcela/Achievement 4'

## products.csv

In [43]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## orders.csv

In [44]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data consistency checks

In [45]:
# Create a dataframe
df_test = pd.DataFrame()

In [46]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [47]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [48]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [49]:
# Converting datatype from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# 04. Finding missing values

In [50]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [51]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [52]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14   12.20
68             69          NaN        26              7   11.80
115           116          NaN        93              3   10.80
261           262          NaN       110             13   12.10
525           525          NaN       109             11    1.20
1511         1511          NaN        84             16   14.30
1780         1780          NaN       126             11   12.30
2240         2240          NaN        52              1   14.20
2586         2586          NaN       104             13   12.40
3159         3159          NaN       126             11   13.10
3230         3230          NaN       120             16   14.40
3736         3736          NaN        41              8   14.80
4283         4283          NaN        77              7   14.40
4790         4790          NaN        91             16   14.50
38187       38183          NaN        39             12   20.90
40444       40440          NaN       120             16   14.80

In [53]:
# Checking the number of rows in product dataframe
df_prods.shape

(49693, 5)

In [54]:
# Creating a new dataframe for non missing value
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [55]:
df_prods_clean.shape

(49677, 5)

In [56]:
# Looking for duplicates in products dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [57]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11    4.80  
18459        27              5    9.20  
26810        27              5   13.40  
35309       121             14    6.80  
35495       127             11    9.90

In [58]:
df_prods_clean.shape

(49677, 5)

In [59]:
# Creating new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [60]:
df_prods_clean_no_dups.shape

(49672, 5)

In [61]:
# Exporting new dataframe as products_checked
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task Steps

## 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [62]:
# Adding code to get rid of exponents
pd.options.display.float_format = '{:.2f}'.format

In [63]:
print(df_ords.columns)

Index(['Unnamed: 0', 'order_id', 'user_id', 'eval_set', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')


In [64]:
# Removing 'Unnamed: 0'
df_ords.drop(columns='Unnamed: 0', inplace=True)

In [65]:
df_ords.describe()

order_id    user_id  order_number  orders_day_of_week  \
count 3421083.00 3421083.00    3421083.00          3421083.00   
mean  1710542.00 1710542.00         17.15                2.78   
std    987581.74  987581.74         17.73                2.05   
min         1.00       1.00          1.00                0.00   
25%    855271.50  855271.50          5.00                1.00   
50%   1710542.00 1710542.00         11.00                3.00   
75%   2565812.50 2565812.50         23.00                5.00   
max   3421083.00 3421083.00        100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

### In my review of the dataset's summary statistics, I found anomalies in the 'days_since_prior_order' column. The minimum is 0 and the maximum, curiously, is 3 days. This conflicts with the 50th percentile value of 7 days, which is higher than the maximum. Similarly, for the 'order_number', both the minimum and maximum are 1, contradicting the 25th percentile of 5. These inconsistencies in the data suggest a need for further investigation to clarify these discrepancies.

## 3. Check for mixed-type data in your df_ords dataframe

In [66]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


## 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type

### No mixed-type data found.

## 5. Run a check for missing values in your df_ords dataframe

In [67]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [68]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [69]:
df_ords_nan

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329  2539329    prior             1                   2   
11        2168274  2168274    prior             1                   2   
26        1374495  1374495    prior             1                   1   
39        3343014  3343014    prior             1                   6   
45        2717275  2717275    prior             1                   3   
...           ...      ...      ...           ...                 ...   
3420930    969311   969311    prior             1                   4   
3420934   3189322  3189322    prior             1                   3   
3421002   2166133  2166133    prior             1                   6   
3421019   2227043  2227043    prior             1                   1   
3421069   3154581  3154581    prior             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### Upon examining the dataset, I discovered that the 206,209 missing values in the 'days_since_prior_order' column correspond to first-time customers. These customers, identified by an order number of "1", logically have no prior order history, hence the absence of data in this column. This pattern aligns with the maximum value in the 'user_id' column, indicating that each new customer's first order accounts for one missing value.

## 6. Address the missing values using an appropriate method

### I've decided to leave the missing values in the 'days_since_prior_order' column unchanged. These values indicate first-time orders and are crucial for understanding new customer behavior. Modifying or removing them could mask this important information. 

## 7. Run a check for duplicate values in your df_ords data

In [70]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [71]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [72]:
df_ords_dups.size

0

### No duplicates found

## 8. Address the duplicates using an appropriate method

### No duplicates found

## 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names

In [73]:
df_ords.shape

(3421083, 7)

In [74]:
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [75]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))